pkg/controllers/node/provisioner_drift.go

```diff
@@ -0,0 +1,33 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/scheduling"
+)
+
+func isProvisionerDrifted(machine *v1alpha5.Machine, provisioner *v1alpha5.Provisioner) bool {
+	nodeRequirementsDrifted := isNodeRequirementsDrifted(machine, provisioner)
+
+	return nodeRequirementsDrifted
+}
+
+func isNodeRequirementsDrifted(machine *v1alpha5.Machine, provisioner *v1alpha5.Provisioner) bool {
+	machineRequirements := scheduling.NewNodeSelectorRequirements(machine.Spec.Requirements...)
```

Comment:
I might be wrong, but I don't know if we can just simply do this. You'll see in the scheduling code that we construct a list of node requirements from the labels, and spinning up my own simple machine from the getting started guide, the list of requirements here don't include most of the labels that we create the machine/node with. 

I think you'll need to try this out (probably best done with more suite tests :) ), but we currently allow non-interesected keys from the left and right side to be allowed. 

Some thoughts for you to consider: 
- What's the best source of truth in the cluster to check if a machine is within provisioner requirements?
- Let's say this source of truth isn't perfect. How do we treat nodes/machines that have imperfect information?
- How should non-intersecting requirements keys work?